module/tools/OdometryBenchmark/src/OdometryBenchmark.cpp

```diff
@@ -0,0 +1,179 @@
+    using message::nbs::player::PlaybackState;
+    using message::nbs::player::PlayRequest;
+    using message::nbs::player::SetModeRequest;
+    using message::nbs::player::PlaybackMode::FAST;
+    using message::nbs::player::PlaybackMode::REALTIME;
+    using message::nbs::player::PlaybackMode::SEQUENTIAL;
+
+    using message::input::Sensors;
+    using message::platform::RawSensors;
+
+    using NUClear::message::CommandLineArguments;
+
+    OdometryBenchmark::OdometryBenchmark(std::unique_ptr<NUClear::Environment> environment)
+        : Reactor(std::move(environment)), config{} {
+
+        on<Configuration>("OdometryBenchmark.yaml").then([this](const Configuration& cfg) {
+            // Use configuration here from file OdometryBenchmark.yaml
+            this->log_level = cfg["log_level"].as<NUClear::LogLevel>();
+
+            auto playback_mode = cfg["playback_mode"].as<std::string>();
+            if (playback_mode == "FAST") {
+                config.mode = FAST;
+            }
+            else if (playback_mode == "SEQUENTIAL") {
+                config.mode = SEQUENTIAL;
+            }
+            else if (playback_mode == "REALTIME") {
+                config.mode = REALTIME;
+            }
+            else {
+                log<ERROR>("Playback mode is invalid, stopping playback");
+                powerplant.shutdown();
+            }
+
+            // Update which types we will be playing
+            for (const auto& setting : cfg["messages"]) {
+                auto name    = setting.first.as<std::string>();
+                bool enabled = setting.second.as<bool>();
+                if (enabled) {
+                    config.messages.push_back(name);
+                }
+            }
+        });
+
+        on<Startup, With<CommandLineArguments>>().then([this](const CommandLineArguments& args) {
+            // Set playback mode
+            auto set_mode_request  = std::make_unique<SetModeRequest>();
+            set_mode_request->mode = config.mode;
+            emit<Scope::INLINE>(set_mode_request);
+
+            // Delay
+            std::this_thread::sleep_for(std::chrono::milliseconds(500));
+
+            // Load the files
+            auto load_request      = std::make_unique<LoadRequest>();
+            load_request->files    = std::vector<std::string>(std::next(args.begin()), args.end());
+            load_request->messages = config.messages;
+            emit<Scope::INLINE>(std::move(load_request));
+
+            // Delay
+            std::this_thread::sleep_for(std::chrono::milliseconds(500));
+
+            // Start playback
+            emit<Scope::INLINE>(std::make_unique<PlayRequest>());
+        });
+
+        on<Trigger<Sensors>, With<RawSensors>>().then([this](const Sensors& sensors, const RawSensors& raw_sensors) {
+            // Check if ground truth odometry data exists
+            if (!raw_sensors.odometry_ground_truth.exists) {
+                return;
+            }
```

Comment:
might be good to have an error unless this is expected to happen sometimes?

Reply:
Good idea. Added.